compiler/src/dmd/dsymbolsem.d

```diff
@@ -1037,8 +1037,12 @@ private extern(C++) final class DsymbolSemanticVisitor : Visitor
                             ex = (cast(AssignExp)ex).e2;
                         if (auto ne = ex.isNewExp())
                         {
-                            // See if initializer is a NewExp that can be allocated on the stack
-                            if (dsym.type.toBasetype().ty == Tclass)
+                            /* See if initializer is a NewExp that can be allocated on the stack.
+                             * Cannot do it if constructor is not `scope` because the `this` can leak.
+                             * https://issues.dlang.org/show_bug.cgi?id=23145
+                             */
+                            if (dsym.type.toBasetype().ty == Tclass &&
+                                (!ne.member || ne.member.storage_class & STC.scope_))
```

Comment:
So the compiler will silently heap allocate if I forget `scope` on my constructor ? It should definitely be an error.